File: typescript/client/convex.ts
import { randomUUID } from "node:crypto";
import WebSocket from "ws";

const MAX_RESPONSE_BYTES = 2 * 1024 * 1024;
const INITIAL_TIMESTAMP = "AAAAAAAAAAA=";
const MAX_PENDING_UPDATES = 16;
export type JsonObject = Record<string, unknown>;
export type CallResult = { value: unknown; logs: string[] };
export type LiveUpdate = {
  value?: unknown;
  logs?: string[];
  error?: ConvexError;
};

export class ConvexError extends Error {
  data?: unknown;
  logs?: string[];
  operation?: string;
  constructor(
    name: string,
    message: string,
    details: Partial<ConvexError> = {},
  ) {
    super(message);
    this.name = name;
    Object.assign(this, details);
  }
}

export class Client {
  private readonly deploymentUrl: string;
  private readonly clientVersion: string;
  private authToken: string;
  private closed = false;
  private live: LiveManager | null = null;
  constructor(
    deploymentUrl: string,
    options: { clientVersion?: string; authToken?: string } = {},
  ) {
    const url = new URL(deploymentUrl);
    if (
      !["http:", "https:"].includes(url.protocol) ||
      url.username ||
      url.password
    )
      throw new TypeError(
        "Convex deployment URL must be an absolute HTTP(S) URL",
      );
    url.search = "";
    url.hash = "";
    url.pathname = url.pathname.replace(/\/$/, "");
    this.deploymentUrl = url.toString().replace(/\/$/, "");
    this.clientVersion = options.clientVersion ?? "typescript-0.1.0";
    this.authToken = options.authToken ?? "";
  }
  setAuth(token: string): void {
    if (this.closed) throw new ConvexError("ClosedError", "client is closed");
    this.authToken = token;
  }
  query(path: string, args: JsonObject = {}): Promise<CallResult> {
    return this.call("query", path, args);
  }
  mutation(path: string, args: JsonObject = {}): Promise<CallResult> {
    return this.call("mutation", path, args);
  }
  action(path: string, args: JsonObject = {}): Promise<CallResult> {
    return this.call("action", path, args);
  }
  private async call(
    operation: string,
    path: string,
    args: JsonObject,
  ): Promise<CallResult> {
    if (!path) throw new TypeError("Convex function path is required");
    if (!args || Array.isArray(args) || typeof args !== "object")
      throw new TypeError("Convex arguments must be a named JSON object");
    if (this.closed) throw new ConvexError("ClosedError", "client is closed");
    let response: Response;
    try {
      response = await fetch(`${this.deploymentUrl}/api/${operation}`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          accept: "application/json",
          "Convex-Client": this.clientVersion,
          ...(this.authToken
            ? { authorization: `Bearer ${this.authToken}` }
            : {}),
        },
        body: JSON.stringify({ path, args, format: "json" }),
      });
    } catch (cause) {
      throw new ConvexError(
        "TransportError",
        `${operation} transport failed: ${String(cause)}`,
      );
    }
    const text = await response.text();
    if (Buffer.byteLength(text) > MAX_RESPONSE_BYTES)
      throw new ConvexError(
        "TransportError",
        `${operation} response exceeds ${MAX_RESPONSE_BYTES} bytes`,
      );
    let body: Record<string, unknown>;
    try {
      body = JSON.parse(text) as Record<string, unknown>;
    } catch {
      throw new ConvexError(
        "TransportError",
        `HTTP ${response.status} returned non-Convex JSON`,
      );
    }
    if (body.status === "success" && Object.hasOwn(body, "value"))
      return {
        value: body.value,
        logs: (body.logLines as string[] | undefined) ?? [],
      };
    if (body.status === "error")
      throw new ConvexError(
        "FunctionError",
        String(body.errorMessage ?? "Convex function failed"),
        {
          data: body.errorData,
          logs: (body.logLines as string[] | undefined) ?? [],
          operation,
        },
      );
    throw new ConvexError(
      "ProtocolError",
      `HTTP ${response.status} response has unknown status ${JSON.stringify(body.status)}`,
    );
  }
  subscribe(path: string, args: JsonObject = {}): Subscription {
    if (this.closed) throw new ConvexError("ClosedError", "client is closed");
    this.live ??= new LiveManager(this.deploymentUrl, this.clientVersion);
    return this.live.subscribe(path, args);
  }
  async debugDisconnectForAdapter(): Promise<void> {
    if (!this.live)
      throw new ConvexError(
        "ProtocolError",
        "Live WebSocket has not been started",
      );
    this.live.debugDisconnect();
  }
  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    await this.live?.close();
  }
}

type Version = { querySet: number; identity: number; ts: string };
type Modification = {
  type: string;
  queryId: number;
  value?: unknown;
  logLines?: string[];
  errorMessage?: string;
  errorData?: unknown;
};
class LiveManager {
  private readonly url: string;
  private readonly clientVersion: string;
  private subscriptions = new Map<number, Subscription>();
  private nextId = 0;
  private querySetVersion = 0;
  private remoteVersion: Version = zeroVersion();
  private connectionCount = 0;
  private lastCloseReason = "InitialConnect";
  private closed = false;
  private socket: WebSocket | null = null;
  private reconnectDelay = 100;
  private connecting = false;
  private reconnectTimer: NodeJS.Timeout | null = null;
  constructor(deploymentUrl: string, clientVersion: string) {
    const url = new URL(deploymentUrl);
    url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
    url.pathname = `${url.pathname.replace(/\/$/, "")}/api/sync`;
    this.url = url.toString();
    this.clientVersion = clientVersion;
  }
  subscribe(path: string, args: JsonObject): Subscription {
    if (!path || !args || Array.isArray(args) || typeof args !== "object")
      throw new TypeError(
        "Live query requires a path and named JSON arguments",
      );
    const subscription = new Subscription(this, this.nextId++, path, args);
    this.subscriptions.set(subscription.id, subscription);
    this.ensureConnection();
    if (this.socket?.readyState === WebSocket.OPEN)
      this.modify([
        { type: "Add", queryId: subscription.id, udfPath: path, args: [args] },
      ]);
    return subscription;
  }
  unsubscribe(subscription: Subscription): void {
    if (!this.subscriptions.delete(subscription.id)) return;
    if (this.socket?.readyState === WebSocket.OPEN)
      this.modify([{ type: "Remove", queryId: subscription.id }]);
  }
  debugDisconnect(): void {
    if (this.socket?.readyState !== WebSocket.OPEN)
      throw new ConvexError(
        "TransportError",
        "Live WebSocket is not connected",
      );
    this.socket.terminate();
  }
  async close(): Promise<void> {
    this.closed = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    for (const subscription of this.subscriptions.values())
      subscription.finish();
    this.subscriptions.clear();
    const socket = this.socket;
    if (!socket || socket.readyState === WebSocket.CLOSED) return;
    await new Promise<void>((resolve) => {
      socket.once("close", () => resolve());
      socket.close();
    });
  }
  private ensureConnection(): void {
    if (
      this.closed ||
      !this.subscriptions.size ||
      this.connecting ||
      (this.socket && this.socket.readyState <= WebSocket.OPEN)
    )
      return;
    this.connecting = true;
    this.socket = new WebSocket(this.url, {
      headers: { "Convex-Client": this.clientVersion },
    });
    this.socket.on("open", () => {
      this.connecting = false;
      this.reconnectDelay = 100;
      this.querySetVersion = 0;
      this.remoteVersion = zeroVersion();
      this.socket?.send(
        JSON.stringify({
          type: "Connect",
          sessionId: randomUUID().replaceAll("-", ""),
          connectionCount: this.connectionCount,
          lastCloseReason: this.lastCloseReason,
          clientTs: 0,
        }),
      );
      const modifications = [...this.subscriptions.values()].map((s) => ({
        type: "Add",
        queryId: s.id,
        udfPath: s.path,
        args: [s.args],
      }));
      if (modifications.length) this.modify(modifications);
    });
    this.socket.on("message", (data) => this.message(data.toString()));
    this.socket.on("error", (error) => {
      this.lastCloseReason = error.message;
    });
    this.socket.on("close", () => {
      this.connecting = false;
      this.connectionCount++;
      this.socket = null;
      if (!this.closed && this.subscriptions.size) {
        const delay = this.reconnectDelay;
        this.reconnectDelay = Math.min(this.reconnectDelay * 2, 15_000);
        this.reconnectTimer = setTimeout(() => this.ensureConnection(), delay);
        this.reconnectTimer.unref();
      }
    });
  }
  private modify(modifications: unknown[]): void {
    if (this.socket?.readyState !== WebSocket.OPEN) return;
    this.socket.send(
      JSON.stringify({
        type: "ModifyQuerySet",
        baseVersion: this.querySetVersion,
        newVersion: this.querySetVersion + 1,
        modifications,
      }),
    );
    this.querySetVersion++;
  }
  private message(raw: string): void {
    let event: {
      type?: string;
      startVersion?: Version;
      endVersion?: Version;
      modifications?: Modification[];
    };
    try {
      event = JSON.parse(raw);
    } catch {
      this.fail(
        new ConvexError("ProtocolError", "Live server sent invalid JSON"),
      );
      return;
    }
    if (event.type !== "Transition") {
      if (
        ["Ping", "MutationResponse", "ActionResponse"].includes(
          event.type ?? "",
        )
      )
        return;
      this.fail(
        new ConvexError(
          "ProtocolError",
          `unsupported Live server message ${event.type}`,
        ),
      );
      return;
    }
    if (
      JSON.stringify(event.startVersion) !== JSON.stringify(this.remoteVersion)
    ) {
      this.fail(
        new ConvexError(
          "ProtocolError",
          "Live transition start version did not match local state",
        ),
      );
      return;
    }
    this.remoteVersion = event.endVersion!;
    for (const change of event.modifications ?? []) {
      const subscription = this.subscriptions.get(change.queryId);
      if (!subscription) continue;
      if (change.type === "QueryUpdated")
        subscription.push({ value: change.value, logs: change.logLines ?? [] });
      else if (change.type === "QueryFailed")
        subscription.push({
          error: new ConvexError(
            "FunctionError",
            change.errorMessage ?? "Live query failed",
            { data: change.errorData, logs: change.logLines ?? [] },
          ),
        });
      else if (change.type !== "QueryRemoved") {
        this.fail(
          new ConvexError(
            "ProtocolError",
            `unsupported Live modification ${change.type}`,
          ),
        );
        return;
      }
    }
  }
  private fail(error: ConvexError): void {
    for (const subscription of this.subscriptions.values())
      subscription.push({ error });
    this.socket?.terminate();
  }
}
export class Subscription implements AsyncIterable<LiveUpdate> {
  private queue: LiveUpdate[] = [];
  private waiter: ((result: IteratorResult<LiveUpdate>) => void) | null = null;
  private done = false;
  constructor(
    private manager: LiveManager,
    readonly id: number,
    readonly path: string,
    readonly args: JsonObject,
  ) {}
  [Symbol.asyncIterator](): AsyncIterator<LiveUpdate> {
    return { next: () => this.next() };
  }
  next(): Promise<IteratorResult<LiveUpdate>> {
    if (this.queue.length)
      return Promise.resolve({ value: this.queue.shift()!, done: false });
    if (this.done) return Promise.resolve({ value: undefined, done: true });
    return new Promise((resolve) => {
      this.waiter = resolve;
    });
  }
  push(update: LiveUpdate): void {
    if (this.done) return;
    if (this.waiter) {
      const resolve = this.waiter;
      this.waiter = null;
      resolve({ value: update, done: false });
      return;
    }
    if (this.queue.length === MAX_PENDING_UPDATES) this.queue.shift();
    this.queue.push(update);
  }
  finish(): void {
    this.done = true;
    this.waiter?.({ value: undefined, done: true });
    this.waiter = null;
  }
  async close(): Promise<void> {
    if (this.done) return;
    this.finish();
    this.manager.unsubscribe(this);
  }
}
function zeroVersion(): Version {
  return { querySet: 0, identity: 0, ts: INITIAL_TIMESTAMP };
}

File: cpp/client/convex.cpp
#include "convex.hpp"
#include <boost/asio/connect.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/asio/ssl.hpp>
#include <boost/beast/core.hpp>
#include <boost/beast/http.hpp>
#include <boost/beast/ssl.hpp>
#include <openssl/ssl.h>

namespace convex {
namespace asio = boost::asio; namespace beast = boost::beast; namespace http = beast::http; namespace ssl = asio::ssl; using tcp = asio::ip::tcp;
struct Parts { std::string host, port, base; bool secure; };
template <typename Stream> static Result perform_http(Stream& stream, const Parts& p, const std::string& token, const std::string& op, const std::string& path, const Json& args) {
  http::request<http::string_body> req{http::verb::post, p.base + "/api/" + op, 11}; req.set(http::field::host, p.host); req.set(http::field::content_type, "application/json"); req.set(http::field::accept, "application/json"); req.set("Convex-Client", "cpp-0.1.0"); if (!token.empty()) req.set(http::field::authorization, "Bearer " + token); req.body() = Json{{"path",path},{"args",args},{"format","json"}}.dump(); req.prepare_payload(); http::write(stream, req);
  beast::flat_buffer buffer; http::response<http::string_body> response; http::read(stream, buffer, response); if (response.body().size() > 2 * 1024 * 1024) throw Error("response exceeds 2097152 bytes"); Json decoded; try { decoded = Json::parse(response.body()); } catch (...) { throw Error("HTTP response was not a Convex response"); } auto logs = decoded.value("logLines", std::vector<std::string>{}); if (decoded.value("status", "") == "success") { if (!decoded.contains("value")) throw Error("success response omitted value"); return {decoded["value"], logs}; } if (decoded.value("status", "") == "error") throw FunctionError(decoded.value("errorMessage", "Convex function failed"), decoded.value("errorData", Json()), logs); throw Error("HTTP response has unknown Convex status");
}
static Parts parse_url(const std::string& value) {
  auto scheme = value.find("://"); if (scheme == std::string::npos) throw Error("Convex deployment URL must use http or https");
  Parts p; auto name = value.substr(0, scheme); p.secure = name == "https"; if (!p.secure && name != "http") throw Error("Convex deployment URL must use http or https");
  const auto start = scheme + 3;
  const auto slash = value.find('/', start);
  const auto authority = value.substr(start, slash - start);
  if (authority.empty() || authority.find('@') != std::string::npos) throw Error("Convex deployment URL must include a host and not user information");
  auto colon = authority.rfind(':'); p.host = colon == std::string::npos ? authority : authority.substr(0, colon); p.port = colon == std::string::npos ? (p.secure ? "443" : "80") : authority.substr(colon + 1); p.base = slash == std::string::npos ? "" : value.substr(slash); while (!p.base.empty() && p.base.back() == '/') p.base.pop_back(); return p;
}
Client::Client(std::string url, std::string token) : url_(std::move(url)), token_(std::move(token)) { parse_url(url_); }
void Client::set_auth(std::string token) { if (closed_) throw Error("Convex client is closed"); token_ = std::move(token); }
void Client::close() { closed_ = true; }
void Client::debug_disconnect_for_adapter() { if (closed_) throw Error("Convex client is closed"); bool disconnected = false; for (auto it = subscriptions_.begin(); it != subscriptions_.end();) { if (auto subscription = it->lock()) { subscription->debug_disconnect(); disconnected = true; ++it; } else it = subscriptions_.erase(it); } if (!disconnected) throw Error("Live WebSocket is not connected"); }
Result Client::query(const std::string& path, const Json& args) { return call("query", path, args); }
Result Client::mutation(const std::string& path, const Json& args) { return call("mutation", path, args); }
Result Client::action(const std::string& path, const Json& args) { return call("action", path, args); }
std::shared_ptr<Subscription> Client::subscribe(const std::string& path, const Json& args) { if (closed_) throw Error("Convex client is closed"); if (path.empty() || !args.is_object()) throw Error("Live query requires a path and named JSON arguments"); auto subscription = std::shared_ptr<Subscription>(new Subscription(url_, path, args)); subscriptions_.push_back(subscription); return subscription; }
Result Client::call(const std::string& op, const std::string& path, const Json& args) {
  if (closed_) throw Error("Convex client is closed");
  if (path.empty()) throw Error("Convex function path is required");
  if (!args.is_object()) throw Error("Convex arguments must be a named JSON object");
  auto p = parse_url(url_);
  asio::io_context io; tcp::resolver resolver(io); auto endpoints = resolver.resolve(p.host, p.port);
  if (!p.secure) { beast::tcp_stream stream(io); stream.connect(endpoints); return perform_http(stream, p, token_, op, path, args); }
  ssl::context ctx(ssl::context::tls_client); ctx.set_default_verify_paths(); ctx.set_verify_mode(ssl::verify_peer); beast::ssl_stream<beast::tcp_stream> stream(io, ctx); stream.set_verify_callback(ssl::host_name_verification(p.host));
  if (!SSL_set_tlsext_host_name(stream.native_handle(), p.host.c_str())) throw Error("configure TLS server name");
  beast::get_lowest_layer(stream).connect(endpoints);
  stream.handshake(ssl::stream_base::client);
  auto result = perform_http(stream, p, token_, op, path, args);
  beast::error_code ignored; stream.shutdown(ignored); return result;
}
}
